mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala

```diff
@@ -169,12 +182,29 @@ object KMeansModel extends MLReadable[KMeansModel] {
 
   /** [[MLWriter]] instance for [[KMeansModel]] */
   private[KMeansModel] class KMeansModelWriter(instance: KMeansModel) extends MLWriter {
+    import org.json4s.JsonDSL._
 
     private case class Data(clusterCenters: Array[Vector])
 
     override protected def saveImpl(path: String): Unit = {
-      // Save metadata and Params
-      DefaultParamsWriter.saveMetadata(instance, path, sc)
+      if (instance.isSet(instance.initialModel)) {
+        val initialModelPath = new Path(path, "initial-model").toString
+        val initialModel = instance.getInitialModel
+        initialModel.save(initialModelPath)
+
+        // Remove the initialModel temporarily
+        instance.clear(instance.initialModel)
```

Comment:
It's probably not a good idea for this serialization method to modify the model. Two potential problem scenarios come to mind: (a) The call to saveMetadata() below fails, leaving the entire KMeansModel object in an inconsistent state; or (b) another thread could be accessing the initialModel field while the current thread calls saveImpl()